hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/ChunkManagerFactory.java

```diff
@@ -0,0 +1,89 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.keyvalue.impl;
+
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hdds.HddsConfigKeys;
+import org.apache.hadoop.ozone.container.keyvalue.interfaces.ChunkManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_CONTAINER_PERSISTDATA;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_CONTAINER_PERSISTDATA_DEFAULT;
+
+/**
+ * Select an appropriate ChunkManager implementation as per config setting.
+ * Ozone ChunkManager is a Singleton
+ */
+public final class ChunkManagerFactory {
+  static final Logger LOG = LoggerFactory.getLogger(ChunkManagerFactory.class);
+
+  private static ChunkManager instance = null;
+  private static boolean syncChunks = false;
+
+  private ChunkManagerFactory() {
+  }
+
+  public static ChunkManager getChunkManager(Configuration config,
+      boolean sync) {
+    if (instance == null) {
+      synchronized (ChunkManagerFactory.class) {
+        if (instance == null) {
+          instance = createChunkManager(config, sync);
+          syncChunks = sync;
+        }
+      }
+    }
+
+    Preconditions.checkArgument((syncChunks == sync),
+        "value of sync conflicts with previous invocation");
+    return instance;
+  }
+
+  private static ChunkManager createChunkManager(Configuration config,
+      boolean sync) {
+    ChunkManager manager = null;
+    boolean persist = config.getBoolean(HDDS_CONTAINER_PERSISTDATA,
+        HDDS_CONTAINER_PERSISTDATA_DEFAULT);
+
+    if (persist == false) {
+      boolean scrubber = config.getBoolean(
+          HddsConfigKeys.HDDS_CONTAINERSCRUB_ENABLED,
+          HddsConfigKeys.HDDS_CONTAINERSCRUB_ENABLED_DEFAULT);
+      if (scrubber) {
+        // Data Scrubber needs to be disabled for non-persistent chunks.
+        LOG.warn("Failed to set " + HDDS_CONTAINER_PERSISTDATA + " to false."
+            + " Please set " + HddsConfigKeys.HDDS_CONTAINERSCRUB_ENABLED
+            + " also to false to enable non-persistent containers.");
+        persist = true;
+      }
+    }
+
+    if (persist == true) {
+      manager = new ChunkManagerImpl(sync);
+    } else {
+      LOG.warn(HDDS_CONTAINER_PERSISTDATA
```

Comment:
Also augment this message to say that if this setting should never be enabled outside of a test environment.